src/main/scala/vexiiriscv/execute/IterativeShiftPlugin.scala

```diff
@@ -0,0 +1,201 @@
+      rightShifts.sorted.foreach(n => {
+        val doIt = if(n > 1) amplitude >= n else True
+        val input = (((ARITHMETIC && srcp.SRC1.msb) #* n) ## shiftReg) >> n
+        val ampl = amplitude - n
+        muxInputs.append((doIt, input, ampl))
+      })
+
+      leftShifts.sorted.foreach(n => {
+        // mask away bits shifted into upper 32 bit in case of SLLW instruction
+        val mask_w = if (Riscv.XLEN.get == 32) {
+          True #* 32
+        } else {
+          (True #* (32 - n)) ## (!IS_W #* n) ## (True #* 32)
+        }
+        val doIt = LEFT & (if(n > 1) amplitude >= n else True)
+        val input = (shiftReg |<< n) & mask_w
+        val ampl = amplitude - n
+        muxInputs.append((doIt, input, ampl))
+      })
+
+      // we only need flip if there is no left shift
+      // keep track of flip indices so that we can skip them when needed
+      val flipIdxs = mutable.ArrayBuffer[Int]()
+      if (leftShifts.isEmpty) {
+        val doFlip = LEFT & ((busy & !flipped) | (if (lateResult) (busy & amplitude === 0) else done))
+
+        flipIdxs.append(muxInputs.size)
+        muxInputs.append((doFlip, shiftReg.reversed, amplitude))
+        if (Riscv.XLEN.get == 64) {
+          flipIdxs.append(muxInputs.size)
+          muxInputs.append((IS_W & doFlip, False #* 32 ## shiftReg(31 downto 0).reversed, amplitude))
+        }
+      }
+      muxInputs.append((selected & !busy, signExtended, shamt))
+
+      val selector = OHMasking.last(Cat(muxInputs.map(_._1)))
+      val muxed = MuxOH(selector, muxInputs.map(_._2))
+      shiftReg := muxed
+
+      // if we use flip(s), then we don't need them in the amplitude MUX, we just just not
+      // EN the amplitude register in that case, saving a few gates
+      val updatedAmplitude = MuxOH(
+        Cat(selector.asBools.zipWithIndex.flatMap{case (b, n) => if(flipIdxs.contains(n)) None else Some(b)}),
+        muxInputs.zipWithIndex.flatMap{case(i, n) => if(flipIdxs.contains(n)) None else Some(i._3)}
+      )
+
+      val anyFlip = flipIdxs.foldLeft(False){case (r, n) => r | muxInputs(n)._1}
+      if(flipIdxs.nonEmpty) {
+        when(anyFlip) {
+          flipped := !flipped
+        }
+      }
+      when(!anyFlip) {
+        amplitude := updatedAmplitude
+      }
+
+      // if we do the load, initialize some other parts of the state as well
+      when(selector.msb) {
+        flipped := False  
+        busy := (if(!lateResult) shamt =/= 0 else True)
+      }
+
+      when((busy && done) || unscheduleRequest) {
+        busy := False
+      }
+
+      val freeze = selected && !done && !unscheduleRequest
+      el.freezeWhen(freeze)
+
+      SHIFT_RESULT := (if (lateResult) shiftReg else muxed)
+    }
+
+    val format = new el.Execute(formatAt) {
+      wb.valid := SEL
+      wb.payload := SHIFT_RESULT
+      if(Riscv.XLEN.get == 64) {
+        when(ARITHMETIC & IS_W) {
+          wb.payload(32, 32 bit).setAllTo(SHIFT_RESULT(31))
```

Comment:
This seems redondant with the `ifp.signExtend(wb, layer(op), 32)` 

ifp => IntFormatPlugin => a plugin which provide a integer writeback path with optional static sign/zero extends

Reply:
right, done^